dolphinscheduler-listener/dolphinscheduler-listener-service/src/main/java/org/apache/dolphinscheduler/listener/service/ListenerEventPublishService.java

```diff
@@ -0,0 +1,273 @@
+ *  * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ *  * KIND, either express or implied.  See the License for the
+ *  * specific language governing permissions and limitations
+ *  * under the License.
+ *
+ *
+ */
+
+package org.apache.dolphinscheduler.listener.service;
+
+import org.apache.dolphinscheduler.dao.entity.ProcessDefinition;
+import org.apache.dolphinscheduler.dao.entity.ProcessInstance;
+import org.apache.dolphinscheduler.dao.entity.ProcessTaskRelationLog;
+import org.apache.dolphinscheduler.dao.entity.Project;
+import org.apache.dolphinscheduler.dao.entity.ProjectUser;
+import org.apache.dolphinscheduler.dao.entity.TaskDefinition;
+import org.apache.dolphinscheduler.dao.entity.TaskDefinitionLog;
+import org.apache.dolphinscheduler.dao.entity.TaskInstance;
+import org.apache.dolphinscheduler.listener.enums.ListenerEventType;
+import org.apache.dolphinscheduler.listener.event.ListenerEvent;
+import org.apache.dolphinscheduler.listener.event.TaskCreateListenerEvent;
+import org.apache.dolphinscheduler.listener.event.TaskEndListenerEvent;
+import org.apache.dolphinscheduler.listener.event.TaskFailListenerEvent;
+import org.apache.dolphinscheduler.listener.event.TaskRemoveListenerEvent;
+import org.apache.dolphinscheduler.listener.event.TaskStartListenerEvent;
+import org.apache.dolphinscheduler.listener.event.TaskUpdateListenerEvent;
+import org.apache.dolphinscheduler.listener.event.WorkflowCreateListenerEvent;
+import org.apache.dolphinscheduler.listener.event.WorkflowEndListenerEvent;
+import org.apache.dolphinscheduler.listener.event.WorkflowFailListenerEvent;
+import org.apache.dolphinscheduler.listener.event.WorkflowRemoveListenerEvent;
+import org.apache.dolphinscheduler.listener.event.WorkflowStartListenerEvent;
+import org.apache.dolphinscheduler.listener.event.WorkflowUpdateListenerEvent;
+
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import javax.annotation.PostConstruct;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+@Slf4j
+public class ListenerEventPublishService {
+
+    private final BlockingQueue<ListenerEvent> listenerEventQueue = new LinkedBlockingQueue<>();
+
+    @Autowired
+    private ListenerEventProducer producer;
+
+    /**
+     * create a daemon thread to process the listener event queue
+     */
+    @PostConstruct
+    private void init() {
+        Thread thread = new Thread(this::doPublish);
+        thread.setDaemon(true);
+        thread.setName("Listener-Event-Produce-Thread");
+        thread.start();
+    }
+
+    public void publish(ListenerEvent listenerEvent) {
+        if (!listenerEventQueue.offer(listenerEvent)) {
+            log.error("Publish listener event failed, message:{}", listenerEvent);
+        }
+    }
```

Comment:
If the downstream producer crash, this will cause master OOM, this is not reasonable, DolphinScheduler is use to schedule and trigger workflow, why `event` handle will affect its available?